frontend/lib/src/components/widgets/ButtonGroup/ButtonGroup.tsx

```diff
@@ -0,0 +1,321 @@
+    return [...currentSelection, index]
+  }
+  return currentSelection.filter(value => value !== index)
+}
+
+function handleSelection(
+  mode: ButtonGroupProto.ClickMode,
+  index: number,
+  currentSelection?: number[]
+): number[] {
+  if (mode == ButtonGroupProto.ClickMode.MULTI_SELECT) {
+    return handleMultiSelection(index, currentSelection ?? [])
+  }
+  return [index]
+}
+
+function getSingleSelection(currentSelection: number[]): number {
+  if (currentSelection.length === 0) {
+    return -1
+  }
+  return currentSelection[0]
+}
+
+function syncWithWidgetManager(
+  selected: number[],
+  element: ButtonGroupProto,
+  widgetMgr: WidgetStateManager,
+  fragmentId?: string,
+  fromUi = true
+): void {
+  widgetMgr.setIntArrayValue(element, selected, { fromUi: fromUi }, fragmentId)
+}
+
+function getContentElement(content: string): ReactElement {
+  const fontSize = "lg"
+  if (isMaterialIcon(content)) {
+    return <DynamicIcon size={fontSize} iconValue={content} />
+  }
+
+  return (
+    <StreamlitMarkdown
+      source={content}
+      allowHTML={true}
+      style={{
+        marginBottom: 0,
+        width: iconSizes[fontSize],
+        display: "inline-flex",
+      }}
+    />
+  )
+}
+
+/**
+ * Returns true if the element should be shown as selected (even though its technically not).
+ * This is used, for example, to show all elements as selected that come before the actually selected element.
+ *
+ * @param selectionVisualization sets the visualization mode
+ * @param clickMode either SINGLE_SELECT or MULTI_SELECT
+ * @param selected list of selected indices. Since only SINGLE_SELECT is considered, this list will always have a length of 1.
+ * @param index of the current element
+ * @returns true if the element is the selected one, or if click_mode is SINGLE_SELECT and selectionVisualization is set to
+ *  ALL_UP_TO_SELECTED and the index of the element is smaller than the index of the selected element, false otherwise.
+ */
+function showAsSelected(
+  selectionVisualization: ButtonGroupProto.SelectionVisualization,
+  clickMode: ButtonGroupProto.ClickMode,
+  selected: number[],
+  index: number
+): boolean {
+  if (selected.indexOf(index) > -1) {
+    return true
+  }
+
+  if (
+    clickMode !== ButtonGroupProto.ClickMode.SINGLE_SELECT ||
+    selectionVisualization !==
+      ButtonGroupProto.SelectionVisualization.ALL_UP_TO_SELECTED
+  ) {
+    return false
+  }
+
+  return selected.length > 0 && index < selected[0]
+}
+
+function getContent(
+  isVisuallySelected: boolean,
+  fallbackContent: string,
+  selectionContent?: string | undefined | null
+): string {
+  if (isVisuallySelected && selectionContent) {
+    return selectionContent
+  }
+
+  return fallbackContent
+}
+
+function createOptionChild(
+  option: ButtonGroupProto.IOption,
+  index: number,
+  selectionVisualization: ButtonGroupProto.SelectionVisualization,
+  clickMode: ButtonGroupProto.ClickMode,
+  selected: number[]
+  // theme: EmotionTheme
```

Comment:
I feel like I'm rehashing a discussion. Any reason this comment needs to be here?